tests/idris2/operators/operators001/Test.idr

```diff
@@ -0,0 +1,80 @@
+
+module Test
+
+import Data.Vect
+
+typebind infixr 0 =@
+infixr 0 -@
```

Comment:
Is there no hope to get a single operator for the binding & non-binding cases?
It seems really noisy to have to introduce 2 names for each function space.

Reply:
I think there is, but I wasn't ready to introduce more human-level parsing ambiguity. Basically the problem is that if we see `x -!> y`, the parser can already know if it's going to be a binding operator or not. If it is then we could make it behave like `(-!>) x (\_ => y)`. If it's not it would behave like `(-!>) x y`. This is perfectly reasonable and achievable today. But the humans cannot tell the difference between the two, specially since the fixity declaration might come from an import.

So what I suggest is that, because this is a niche feature anyways, we leave the duplication. And once enough people in the community get used to it, and start complaining that it should work both ways, we introduce the ability to with both `(x : ty) -@ y` and `x -@ y`, which should be backward compatible.

Reply:
Given that `->` already works like that, I think having a uniform behaviour
for all function spaces would be less confusing.

Reply:
This is not just for function spaces, specially when you start using `autobind`. Maybe it would make sense for `typebind` to allow the overloading but I still think it's too early to tell if disallowing the overloading is a usability regression or usability improvement. I would still insist on disallowing the overloading for three reasons:
- allowing it is a non-breaking change, we can always add it later
- Allowing binding operators is already a usability improvement, even if it comes with overzealous safety nets
- If the overloading was allowed, I would like to see proper error messaging diagnosing the fact that switching from/to binding syntax is a valid way to resolve the problem. But doing it naively would simply result in an obscure unification error.

Reply:
I think I've already encountered people puzzled by `(x : a) -@ b` not working in the current state of affairs.

Reply:
Indeed, and those users are still getting benefits:
- now they have an alternative with `=@` (assuming we add it to `linear`).
- they get an error message that explains why `-@` does not work.

Like I said in my first response, I am not inherently against implementing the overloading. However I don't think we have enough data about the user experience to make the call right now. Finally, this PR implement #3113 which did not make any claims about overloading, quite the opposite:

> this proposal errs on the side of caution  by making conservative choices with regard to overloading and syntax disambiguation

I'd be happy to discuss the details and implementation of overloading binding feature in a dedicated channel but I think it's out of scope for this PR.

Reply:
I don't like `=@` because I think it should be to `-@` what `=>` is to `->`, again for design uniformity.
But that's a separate bikeshedding discussion.

I am concerned that when we attempt to implement the overloading we will actually discover that
it cannot be done because it makes the grammar too ambiguous and then we'll be stuck with the
current design. Additionally, I find the "not enough data" argument dubious when the choice is between
1. a consistent design across the board
2. a design with more corner cases

Finally, this PR bundles together `typebind` (which is not limited to types IIUC) and
`autobind` (which has nothing to do with auto but is more like an infix notion of let-binding
IIUC) but AFAIU these are orthogonal features. Should we have separate PRs and separate
discussions?

Reply:
> I don't like =@ because I think it should be to -@ what => is to ->, again for design uniformity.
But that's a separate bikeshedding discussion.

Indeed, this PR is about the `autobind` and `typebind` feature, not about `=@` which is not in this diff.

> I am concerned that when we attempt to implement the overloading we will actually discover that
it cannot be done

The error messages can detect when you're misusing a regular operator as a binding operator and vice-versa. Lifting the restriction means changing this error message into the proper desugaring.

> I find the "not enough data" argument dubious 

It comes from the fact that I've been using this diff for my own code for 4 weeks and haven't felt the need to add the overload. Maybe someone else will and I'm waiting for this to happen. In the meantime, I ask that people hold their judgement until they've tried things. As for the two points, I think the design is consistent with the rationale given, and the corner cases are properly handled with helpful error messages.

> AFAIU these are orthogonal features. Should we have separate PRs and separate
discussions

Possibly, but this PR is the result of a proposal written in october 21st, almost 2 months ago. I would have appreciated if those concerned were raised on the appropriate channel during the last 6 weeks. I know we don't have a set timeline for proposal review and implementation (maybe it's time to set one up), but bringing up design concerns, that question the existence of the diff itself, during code review feels bad.

ps: I left 2 design flaws in this PR that I was hoping to get to in the next couple weeks. I'm happy they haven't been noticed yet, but I'm bummed that the alternative (getting the feature scraped) is worse

Reply:
>  I would have appreciated if those concerned were raised on the appropriate channel during the last 6 weeks

I understand that and I am sorry that I did not react earlier but then again
we are all busy and investing time into thoroughly discussing a 4 thousand
words proposal is also a big ask.

I can easily read the test cases using the typebind operators and, although
I still think the design is broken, they behave as I expect. But I must admit
the autobind ones are still pretty confusing / unreadable.

Reply:
I understand this is a big undertaking and I'm happy to provide additional info and context. But the review process is not the right place. We also share an office so I suggest we meet there instead.

> But I must admit the autobind ones are still pretty confusing / unreadable.

I entirely agree with this and this is why I'm am resistant to adding the overloading syntax at the same time. This is already confusing enough as it is. 

I suggest we put this back as a draft, I'll add the missing test case for regular used as typebind, fix the two flaws in the backlog (allow `_` as a name and use levenshtein distance to suggest other operators in diagnostics)